src/controller/proxy/controller.go

```diff
@@ -155,6 +155,9 @@ func (c *controller) UseLocalManifest(ctx context.Context, art lib.ArtifactInfo,
 	remoteRepo := getRemoteRepo(art)
 	exist, desc, err := remote.ManifestExist(remoteRepo, getReference(art)) // HEAD
 	if err != nil {
+		if errors.IsRateLimitError(err) && a != nil { // if rate limit, use local if it exists, otherwise return error
+			return true, nil, nil
+		}
```

Comment:
IMHO, what needs to be refined is in the path when error happens and the cache is expired or the data is not cached.

In this case, we better return the error more explicitly instead of a 404.  I think a better choice here is to return 500 which wraps the root cause (in this case 429), if 500 can't be rendered in a friendly way for mainstream clients, we find a proper error so most clients can handle it.

Reply:
To avoid leaking any server information. if return 500 error to the client, it just sends the "internal server error" to the client, no other useful information.
see
https://github.com/goharbor/harbor/blob/cc4c08b0a1cef086fdb47d90be8f6b32efe170cf/src/lib/http/error.go#L55-L54